### PR TITLE
JW-785: Spawn VM parametrized function

### PR DESCRIPTION
**Pull request by jablonskim**

### CODE REVIEW

File: CIScripts/SpawnVM.ps1 [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        Copy-Item -ToSession $Session -Path "vrouter\vRouter.msi" -Destination C:\Artifacts\
+        Copy-Item -ToSession $Session -Path "vrouter\utilsMSI.msi" -Destination C:\Artifacts\
+        Copy-Item -ToSession $Session -Path "vrouter\*.cer" -Destination C:\Artifacts\ # TODO: Remove after JW-798
+
+        Invoke-Command -Session $Session -ScriptBlock {
+            Write-Host "Installing Contrail vRouter API"
+            pip2 install C:\Artifacts\contrail-vrouter-api-1.0.tar.gz | Out-Null
+
+            Write-Host "Installing Docker driver"
+            Start-Process msiexec.exe -ArgumentList @("/i", "C:\Artifacts\installer.msi", "/quiet") -Wait
```
Comment (m-kostrzewa) [inline CIScripts/SpawnVM.ps1]: I know this is not in scope of this PR, but if you could rename installer.msi to something like driver.msi and utilsMSI.msi to something like utils.msi, that would be great

**Review by m-kostrzewa — DISMISSED** [review]
This is very clean and nice to read, good job

File: CIScripts/SpawnVM.ps1 [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        })
+
+        return $Sessions
+    }
+
+    function Enable-NBLDebugging {
+        Param ([Parameter(Mandatory = $true)] [System.Management.Automation.Runspaces.PSSession] $Session)
+
+        Invoke-Command -Session $Session -ScriptBlock {
+            New-ItemProperty -Path "HKLM:\System\CurrentControlSet\Services\NDIS\Parameters" -Name TrackNblOwner -Value 1 -PropertyType DWORD -Force | Out-Null
```
Comment (mclapinski) [inline CIScripts/SpawnVM.ps1]: A comment here would be nice

File: CIScripts/SpawnVM.ps1 [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    function Initialize-CrashDumpSaving {
+        Param ([Parameter(Mandatory = $true)] [System.Management.Automation.Runspaces.PSSession] $Session,
+               [Parameter(Mandatory = $true)] [string] $DumpFilesLocation,
+               [Parameter(Mandatory = $true)] [string] $DumpFilesBaseName)
+
+        $DumpFilename = $DumpFilesLocation + "\" + $DumpFilesBaseName + "_" + $Session.ComputerName + ".dmp"
+
+        Invoke-Command -Session $Session -ScriptBlock {
```
Comment (mclapinski) [inline CIScripts/SpawnVM.ps1:113]: I think you should only format the disk you created, not all of them. Write todo comment that passing some UUID is needed.

File: CIScripts/SpawnVM.ps1 [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        }
+
+        Write-Host "Copying Docker driver installer"
+        Copy-Item -ToSession $Session -Path "docker_driver\installer.msi" -Destination C:\Artifacts\
+
+        Write-Host "Copying Contrail vRouter API"
+        Copy-Item -ToSession $Session -Path "agent\contrail-vrouter-api-1.0.tar.gz" -Destination C:\Artifacts\
+
+        Write-Host "Copying vtest"
+        Copy-Item -ToSession $Session -Path "vrouter\utils\vtest" -Destination C:\Artifacts\ -Recurse -Force
```
Comment (mclapinski) [inline CIScripts/SpawnVM.ps1]: "copying vtest scenarios"

File: CIScripts/SpawnVM.ps1 [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            Write-Host "Installing Docker driver"
+            Start-Process msiexec.exe -ArgumentList @("/i", "C:\Artifacts\installer.msi", "/quiet") -Wait
+
+            Write-Host "Installing Utils"
+            Start-Process msiexec.exe -ArgumentList @("/i", "C:\Artifacts\utilsMSI.msi", "/quiet") -Wait
+
+            Write-Host "Installing vRouter Extension"
+            Import-Certificate -CertStoreLocation Cert:\LocalMachine\Root\ "C:\Artifacts\vRouter.cer" | Out-Null # TODO: Remove after JW-798
+            Import-Certificate -CertStoreLocation Cert:\LocalMachine\TrustedPublisher\ "C:\Artifacts\vRouter.cer" | Out-Null # TODO: Remove after JW-798
+            Start-Process msiexec.exe -ArgumentList @("/i", "C:\Artifacts\vRouter.msi", "/quiet") -Wait
```
Comment (mclapinski) [inline CIScripts/SpawnVM.ps1:155]: I'd move this before utils

File: CIScripts/SpawnVM.ps1 [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    Initialize-VIServer -PowerCLIScriptPath $PowerCLIScriptPath -VIServerAccessData $VIServerAccessData
+
+    Write-Host "Starting VMs"
+    $VMNames.ForEach({ New-StartedVM -VMName $_ -VMCreationSettings $VMCreationSettings })
+
+    Write-Host "Waiting for VMs to start..."
+    $VMsList = [Collections.Generic.List[String]] $VMNames
+    $Res = Wait-VMs -VMNames $VMsList
+    if ($Res -eq $false) {
+        Write-Host "Waited for too long, aborting."
```
Comment (mclapinski) [inline CIScripts/SpawnVM.ps1]: Consider throwing exception inside Wait_VMs

**Review by mclapinski — CHANGES_REQUESTED** [review]

File: CIScripts/SpawnVM.ps1 (hunk 1 above)
Comment (jablonskim) [reply CIScripts/SpawnVM.ps1]: I've created tasks JW-805 and JW-806 for that

File: CIScripts/SpawnVM.ps1 [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    Write-Host "Connecting to VIServer"
+    Initialize-VIServer -PowerCLIScriptPath $PowerCLIScriptPath -VIServerAccessData $VIServerAccessData
+
+    Write-Host "Starting VMs"
+    $VMNames.ForEach({ New-StartedVM -VMName $_ -VMCreationSettings $VMCreationSettings })
+
+    Write-Host "Waiting for VMs to start..."
+    $VMsList = [Collections.Generic.List[String]] $VMNames
```
Comment (mclapinski) [inline CIScripts/SpawnVM.ps1:181]: Meh

**Review by mclapinski — APPROVED** [review]

**Pull request merged**